### PR TITLE
Buddies feature

### PR DESCRIPTION
**Pull request by dorumd**
When using the product, we've realised it's missing one very handy capability: seeing where are your buddies sitting. In this MR, a new REST API to store / remove / list buddies is added and FE changes to display with a different shade the bookings made by buddies(on search page), adding a new buddy(on search page) and a new page to manage buddies list(view & remove).

Potential follow-up improvements:
* Add tooltip over the booking with the email of buddy which booked it
* Add explanation what each color used on the search view means
* Search by email and add buddies directly on "My buddies" page

Note: in order to avoid weird UX, this feature requires "Show names" flag to be activated.

Screenshot 2024-01-07 at 22 25 41
Screenshot 2024-01-07 at 22 25 52
Screenshot 2024-01-07 at 22 25 59
Screenshot 2024-01-07 at 22 26 07
Screenshot 2024-01-07 at 22 26 20

### CODE REVIEW

File: server/buddy-router.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	CreateBuddyRequest
+}
+
+func (router *BuddyRouter) setupRoutes(s *mux.Router) {
+	s.HandleFunc("/{id}", router.delete).Methods("DELETE")
+	s.HandleFunc("/", router.create).Methods("POST")
+	s.HandleFunc("/", router.getAll).Methods("GET")
+}
+
+func (router *BuddyRouter) getAll(w http.ResponseWriter, r *http.Request) {
```
Comment (dorumd) [inline server/buddy-router.go:33]: The BE is missing tests, to be added.

File: booking-ui/src/styles/Search.css [hunk 2]
```diff
@@ -163,6 +163,10 @@
     background-color: rgba(184, 37, 222, 0.9) !important;
 }
 
+.space-buddy-desk {
+    background-color: rgba(36, 21, 197, 0.9) !important;
```
Comment (dorumd) [inline booking-ui/src/styles/Search.css:167]: I chose this color but not really familiar with accessibility impact, happy to change with another shade / color.

**Comment by dorumd** [comment]
@virtualzone Hello! I believe this feature is now ready for review. Happy to hear your thoughts.

File: server/buddy-router.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+		res = append(res, m)
+	}
+	SendJSON(w, res)
+}
+
+func (router *BuddyRouter) delete(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	e, err := GetBuddyRepository().GetOne(vars["id"])
+	if err != nil {
+		SendUpdated(w)
```
Comment (virtualzone) [inline server/buddy-router.go:52]: I guess SendNotFound() is more suitable here.

File: server/buddy-router.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+		return
+	}
+	if err := GetBuddyRepository().Delete(e); err != nil {
```
Comment (virtualzone) [inline server/buddy-router.go:55]: A check whether the buddy belongs to the request user is necessary here.

**Review by virtualzone — CHANGES_REQUESTED** [review]
Hi @dorumd, thanks for this great enhancement! Please take a look at my comments in the ```buddy-router.go``` file.

File: server/buddy-router.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+		SendInternalServerError(w)
+		return
+	}
+	SendUpdated(w)
+}
+
+func (router *BuddyRouter) create(w http.ResponseWriter, r *http.Request) {
```
[the remark this answers is not in the document]
Comment (dorumd) [reply server/buddy-router.go:62]: In this case we have a create of a "Buddy" resource, why do you think an update & PUT is necessary?

**Comment by dorumd** [comment]
Thanks for the feedback! @virtualzone Replied & fixed where required. I also noticed the version for go I used to test was newer and had slices as part of standard library, so I replaced it with a helper function instead in `main_test.go`.

**Review by virtualzone — APPROVED** [review]

**Pull request merged**